# apps with good revenue!
 
**in this very first project:**

Our `aim` is to find mobile app profiles that are profitable for the App Store and Google Play markets. We're working as data analysts for a company that builds Android and iOS mobile apps, and our job is to enable our team of developers to make data-driven decisions with respect to the kind of apps they build.

At our company, we only build apps that are **free** to download and install, and our main `source of revenue` consists of in-app ads. This means that our revenue for any given app is mostly influenced by `the number of users that use our app`. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.

at first we will analyze and clean all the data and the data for us here is google play apps at the year of 2018 click [here](https://www.kaggle.com/lava18/google-play-store-apps) to download the data set.

In [1]:
# first we will define the function we will use to extract the data from the file

def explore(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

# import the module then read the data

from csv import reader
apple_store = list(reader(open('D:\Downloads\AppleStore.csv')))
google_store = list(reader(open('D:\Downloads\googleplaystore.csv')))

print(explore(google_store,0,2,True))
print('\n')
print(explore(apple_store,0,2,True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13
None


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 7198
Number of columns: 16
None


We see that the Google Play data set has 10841 apps and 13 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

and our analysis is devide into step first one is:


# 1) cleaning phase:

every data analysis need a clean data-set, because when there is no mistakes or gaps or dublicates in the data-set our analsys will be more precise, we will clean it step by step.

## a) cleaning irregular rows


In [2]:
# now we check whether all the data points are there and no rows in our data is missing for a value

header_row = len(google_store[0])

for row in google_store[1:]:
    index_num = google_store.index(row)
    
    if len(row) != header_row : # we made it with if statement to make sure we won't worry if we run this code more than once 
        del google_store[index_num]
        
# we did not do it for apple store because it already has regular rows

## b) delete the dublicates

now we will check wether there is some dublicates apps in our data or not by runnig some specific test:

In [3]:
for app in google_store:
    name = app[0]
    if name == 'Instagram':
        print(app)
        print('\n')

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




**as we can see there is mulitple dublicates for the same app for that we will:**

- create dictionary to put how much the app occurred in our data
- and also create a list of lists for unique apps

In [4]:
dublicated_apps_dict = {}
unique_apps= []

for row in google_store[1:]:
    app_name = row[0]
    
    if app_name in unique_apps:
        if app_name in dublicated_apps_dict:
            dublicated_apps_dict[app_name]+=1
        else:
            dublicated_apps_dict[app_name]=1
    else:
        unique_apps.append(app_name)
        
# after that we will convert the dictionary to a list of tubles that have the key of the dictionary and the number of times

dublicated_apps_list = []

for key, value in dublicated_apps_dict.items():
    the_tuble =(key,value)
    dublicated_apps_list.append(the_tuble)
    
# finally we will show to the user how many apps that dublicated and the number of unique apps

print('number of dublicated apps: {}\n'.format(len(dublicated_apps_list)))

print ("And here an example of some of the dublicated apps and how much it's dublications:")

for app in dublicated_apps_list[0:7]:
    a,b =app
    print ('{} : {} time/s'.format(a,b))
    
print('\nnumber of unique apps : {} \n\nand the examples from it :\n{}'.format(
    len(unique_apps),'\n'.join(unique_apps[:5])))

number of dublicated apps: 798

And here an example of some of the dublicated apps and how much it's dublications:
Quick PDF Scanner + OCR FREE : 2 time/s
Box : 2 time/s
Google My Business : 2 time/s
ZOOM Cloud Meetings : 1 time/s
join.me - Simple Meetings : 2 time/s
Zenefits : 1 time/s
Google Ads : 2 time/s

number of unique apps : 9659 

and the examples from it :
Photo Editor & Candy Camera & Grid & ScrapBook
Coloring book moana
U Launcher Lite – FREE Live Cool Themes, Hide Apps
Sketch - Draw & Paint
Pixel Draw - Number Art Coloring Book


**after we found how many dublicates are there in our data we need to remove it:**

- we will make dictionary with the ubdated information

- our `criterion` is the highest number of reviews in an app because : `thats mean it is the most recent data should be`

- we will take the apps with the ubdated information

In [5]:
# we will start by creating dictionary with highest number in reviews counts

reviews_max={}

for app in google_store[1:]:
    name = app[0]
    num_reviews = float(app[3])
    
    if name in reviews_max and num_reviews > reviews_max[name] : #
        reviews_max[name]= num_reviews
        
    if name not in reviews_max :
        reviews_max[name] = num_reviews
        
# now we make sure that the dictionay is correc by compare it with number of unique apps that we calculated earlier

if len(reviews_max) == len(unique_apps) :
    print('expected length: {}'.format(len(unique_apps)))
    print('actuall length: {}'.format(len(reviews_max)))

expected length: 9659
actuall length: 9659


**our dictionay that has names as a key and review counts as the value is well sorted.**

- now we will loop throght google play store data set and clean it
- and also will initial two empty lists that stores our new and clean data :

In [6]:
google_store_clean = []
already_added = []

for app in google_store[1:] :
        name = app[0]
        num_reviews = float(app[3])
        
        if num_reviews == reviews_max[name] and name not in already_added :
            google_store_clean.append(app)
            already_added.append(name)

now lets explore the now data:

In [7]:
explore(google_store_clean,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have 9,659 rows, just as expecte, as we cleaned the data points we will go to the next step.

## C) isolate non english apps

**and thats is isolate the none english apps, because we just want the audience to be an english speakers :**

- first of all we will define a `function`, that takes the name of the app and declares if that app name is 'none english' or Not, of course we will take the english one. 

In [8]:
def is_english(string):
    for charc in string :
        if ord(charc) > 127:
            return False
    return True

now lets test the the functon we made:

In [9]:
print(is_english('instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


**as we can see our function is working now lets adjust it alittle bit so we mimimize the impact of lossing to many data points:**

- we can do that if we calculated the number of none english charcters if it less than 3 we will keep the app with us

 

In [10]:
def english_detector_3000(string):
    non_english = 0
    
    for character in string:
        if ord(character) > 127:
            non_english += 1
    
    if non_english > 3:
        return False
    else:
        return True

let's test the new function:

In [11]:
print (english_detector_3000('Docs To Go™ Free Office Suite'))
print (english_detector_3000('Instachat 😜'))
print (english_detector_3000('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


now let's do some filtring :

In [12]:
google_store_english = []
apple_store_english = []

for app in google_store_clean:
    name = app[0]
    if english_detector_3000(name):
        google_store_english.append(app)
        
for app in apple_store:
    name = app[1]  
    if english_detector_3000(name):
        apple_store_english.append(app)
        
explore(google_store_english, 0, 3, True)
print('\n')
explore(apple_store_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9500
Number of columns: 13


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Insta

**as we see we ended up with 9,500 english apps in google store, and 6,101 apps in apple store.**

we mentioned earlier want both english, and free apps now we will isolate the non free apps:

In [13]:
google_store_final = []
apple_store_final =[]

for app in google_store_english:
    
    if app[6] == 'Free':
        google_store_final.append(app)
    
for app in apple_store_english:
    
    if app[4] == '0.0':
        apple_store_final.append(app)
        
print(len(google_store_final))
print('\n')
print(len(apple_store_final))

8759


3169


we ended up now with 8,407 apps from google store and 2,922 apps from apple store and those are good numbers to start Analyse it.


# analysing  Phase :

**As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.**

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

first we will define a function that makes for us a frequence table by genre and category to make the analysing easier .

## a) analyze by genre or category

In [14]:
def frequence_table(data_set, coulmn):
    devide = len(data_set)
    dictionary = {}
    
    for app in data_set :
        name = app[coulmn]
        
        if name in dictionary:
            dictionary[name] += 1
        else :
            dictionary[name] = 1
            
             
    dictionary = {round((value*100/devide), 3): key for key , value in dictionary.items()}
    
    dictionary = sorted(dictionary.items(),reverse = True)
    
    for order ,pair in enumerate(dictionary, start = 1) :
        print(f'#{order} genre is {pair[1]} with : {pair[0]}%\n')

## part one:
## apple store

In [15]:
frequence_table(apple_store_final,11)

#1 genre is Games with : 58.536%

#2 genre is Entertainment with : 7.826%

#3 genre is Photo & Video with : 5.049%

#4 genre is Education with : 3.724%

#5 genre is Social Networking with : 3.282%

#6 genre is Shopping with : 2.524%

#7 genre is Utilities with : 2.398%

#8 genre is Sports with : 2.177%

#9 genre is Music with : 2.051%

#10 genre is Health & Fitness with : 1.988%

#11 genre is Productivity with : 1.704%

#12 genre is Lifestyle with : 1.546%

#13 genre is News with : 1.325%

#14 genre is Travel with : 1.136%

#15 genre is Finance with : 1.104%

#16 genre is Weather with : 0.852%

#17 genre is Food & Drink with : 0.82%

#18 genre is Business with : 0.536%

#19 genre is Book with : 0.379%

#20 genre is Medical with : 0.189%

#21 genre is Catalogs with : 0.126%



we see that, more than the half of our DATA is dominated by the genre games with 59%, and afer that there the runner-up entertaiment with just 7.5% then, with third place comes photo & video editor with 5.1% .

from that we can say there some preference to genres with entertaining purposes like :
(games, social media,  photo & video editors,...etc).

the coustemrs tent to the idea having fun with smart phones more than taking education and practical skills from apps.

## part two:
## google store 

In [16]:
frequence_table(google_store_final,1)

#1 genre is FAMILY with : 18.929%

#2 genre is GAME with : 9.659%

#3 genre is TOOLS with : 8.483%

#4 genre is BUSINESS with : 4.647%

#5 genre is PRODUCTIVITY with : 3.939%

#6 genre is LIFESTYLE with : 3.916%

#7 genre is FINANCE with : 3.722%

#8 genre is MEDICAL with : 3.551%

#9 genre is SPORTS with : 3.334%

#10 genre is PERSONALIZATION with : 3.288%

#11 genre is COMMUNICATION with : 3.254%

#12 genre is HEALTH_AND_FITNESS with : 3.094%

#13 genre is PHOTOGRAPHY with : 2.98%

#14 genre is NEWS_AND_MAGAZINES with : 2.809%

#15 genre is SOCIAL with : 2.649%

#16 genre is TRAVEL_AND_LOCAL with : 2.34%

#17 genre is SHOPPING with : 2.249%

#18 genre is BOOKS_AND_REFERENCE with : 2.146%

#19 genre is DATING with : 1.861%

#20 genre is VIDEO_PLAYERS with : 1.804%

#21 genre is MAPS_AND_NAVIGATION with : 1.381%

#22 genre is FOOD_AND_DRINK with : 1.233%

#23 genre is EDUCATION with : 1.176%

#24 genre is ENTERTAINMENT with : 0.959%

#25 genre is AUTO_AND_VEHICLES with : 0.925%

#26 ge

In [17]:
frequence_table(google_store_final,9)

#1 genre is Tools with : 8.471%

#2 genre is Entertainment with : 6.085%

#3 genre is Education with : 5.389%

#4 genre is Business with : 4.647%

#5 genre is Productivity with : 3.939%

#6 genre is Lifestyle with : 3.905%

#7 genre is Finance with : 3.722%

#8 genre is Medical with : 3.551%

#9 genre is Sports with : 3.402%

#10 genre is Personalization with : 3.288%

#11 genre is Communication with : 3.254%

#12 genre is Action with : 3.105%

#13 genre is Health & Fitness with : 3.094%

#14 genre is Photography with : 2.98%

#15 genre is News & Magazines with : 2.809%

#16 genre is Social with : 2.649%

#17 genre is Travel & Local with : 2.329%

#18 genre is Shopping with : 2.249%

#19 genre is Books & Reference with : 2.146%

#20 genre is Simulation with : 2.055%

#21 genre is Dating with : 1.861%

#22 genre is Arcade with : 1.827%

#23 genre is Video Players & Editors with : 1.781%

#24 genre is Casual with : 1.735%

#25 genre is Maps & Navigation with : 1.381%

#26 genre is Food &

**and for google store there:**
we can notice there no signefcunt differance between the catagory and the prime genre, but most category with percentage are educational purposes and practical knowledge such as (Tools, Education, Business, PRODUCTIVIT,...etc ).

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

if we investigate more and  more we can it is not all just educational purposes there is some KID's games inside FAMILY category as we can see it here:

[![Image](https://i.imgur.com/ALErqJF.png)](https://play.google.com/store/apps/category/FAMILY?hl=en)



# the summary of genre by percentage :
 
**Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.**



## b) analyze by average number of users


we can get good use of the average because eventually our criterion is the number of users

now lets create some functions to calculate the average number of users:

In [18]:
def make_frequence(data_set, coulmn):
    dictionary = {}
    
    for app in data_set :
        name = app[coulmn]
        
        if name in dictionary:
            dictionary[name] += 1
        else :
            dictionary[name] = 1
    return dictionary

def genre_frequence(data_set,table, coulmn,coulmn2): 
    tubles={}
    
    for genre in table:
        devide = table[genre]
        total = 0
        
        for app in data_set:
            genre_name = app[coulmn]
            
            if genre == genre_name:
                total+= float(app[coulmn2])
                
        genre_users = int (total/devide)
        tubles[genre] = genre_users
        
    tubles =sorted({value:key for key ,value in tubles.items()}.items(), reverse =True)
    
    for order, pair in enumerate(tubles, start = 1):
        print(f'#{order} {pair[1]} : has average of {pair[0]:,} users.\n')

## part one:
## apple store

In [19]:
apple_store_genre = make_frequence(apple_store_final, 11)
genre_frequence(apple_store_final,apple_store_genre,11,5)

#1 Navigation : has average of 86,090 users.

#2 Reference : has average of 79,350 users.

#3 Social Networking : has average of 72,916 users.

#4 Music : has average of 58,205 users.

#5 Weather : has average of 54,215 users.

#6 Book : has average of 46,384 users.

#7 Food & Drink : has average of 33,333 users.

#8 Finance : has average of 32,367 users.

#9 Travel : has average of 31,358 users.

#10 Photo & Video : has average of 28,441 users.

#11 Shopping : has average of 27,816 users.

#12 Health & Fitness : has average of 24,037 users.

#13 Sports : has average of 23,008 users.

#14 Games : has average of 22,985 users.

#15 Productivity : has average of 21,799 users.

#16 News : has average of 21,750 users.

#17 Utilities : has average of 19,900 users.

#18 Lifestyle : has average of 16,739 users.

#19 Entertainment : has average of 14,364 users.

#20 Business : has average of 7,491 users.

#21 Education : has average of 7,003 users.

#22 Catalogs : has average of 4,004 users.

#

we can see that the top the genres by nymber of users is navigation, Refernece and Social Networking but we can't just say it is the most popular genres and it is commercial viable because some of the genres may have some huge apps that influence th average.

for examples in social networking and music down below: 

In [20]:
for app in apple_store_final:
        if app[-5] == 'Social Networking' and float(app[5]) >100_000:
            print (f'{app[1]} with {float(app[5]):,} number of users\n')

Facebook with 2,974,676.0 number of users

Pinterest with 1,061,624.0 number of users

Skype for iPhone with 373,519.0 number of users

Messenger with 351,466.0 number of users

Tumblr with 334,293.0 number of users

WhatsApp Messenger with 287,589.0 number of users

Kik with 260,965.0 number of users

ooVoo – Free Video Call, Text and Voice with 177,501.0 number of users

TextNow - Unlimited Text + Calls with 164,963.0 number of users

Viber Messenger – Text & Call with 164,249.0 number of users

Followers - Social Analytics For Instagram with 112,778.0 number of users



In [21]:
for app in apple_store_final:
        if app[-5] == 'Music' and float(app[5]) >100_000:
            print (f'{app[1]} with {float(app[5]):,} number of users\n')

Pandora - Music & Radio with 1,126,879.0 number of users

Spotify Music with 878,563.0 number of users

Shazam - Discover music, artists, videos & lyrics with 402,925.0 number of users

iHeartRadio – Free Music & Radio Stations with 293,228.0 number of users

SoundCloud - Music & Audio with 135,744.0 number of users

Magic Piano by Smule with 131,695.0 number of users

Smule Sing! with 119,316.0 number of users

TuneIn Radio - MLB NBA Audiobooks Podcasts Music with 110,420.0 number of users

Amazon Music with 106,235.0 number of users



## part two:
## with google store

In [22]:
google_table = make_frequence(google_store_final,1)
tubles={}

for genre in google_table:
    devide = google_table[genre]
    total = 0

    for app in google_store_final:
        genre_name = app[1]
        if genre == genre_name :
            totalapp = app[5].replace('+','')
            total += float(totalapp.replace(',',''))
            
    genre_users = int (total/devide)
    tubles[genre] = genre_users

tubles =sorted({value:key for key ,value in tubles.items()}.items(), reverse =True)

for order, pair in enumerate(tubles, start = 1):
    print(f'#{order} {pair[1]} : has average of {pair[0]:,} users.\n')

#1 COMMUNICATION : has average of 38,550,548 users.

#2 VIDEO_PLAYERS : has average of 24,878,048 users.

#3 SOCIAL : has average of 23,628,689 users.

#4 PHOTOGRAPHY : has average of 17,840,110 users.

#5 PRODUCTIVITY : has average of 16,787,331 users.

#6 GAME : has average of 15,571,586 users.

#7 TRAVEL_AND_LOCAL : has average of 14,120,454 users.

#8 ENTERTAINMENT : has average of 11,767,380 users.

#9 TOOLS : has average of 10,902,378 users.

#10 NEWS_AND_MAGAZINES : has average of 9,626,407 users.

#11 BOOKS_AND_REFERENCE : has average of 8,329,168 users.

#12 SHOPPING : has average of 7,103,190 users.

#13 PERSONALIZATION : has average of 5,240,358 users.

#14 WEATHER : has average of 5,212,877 users.

#15 HEALTH_AND_FITNESS : has average of 4,219,697 users.

#16 MAPS_AND_NAVIGATION : has average of 4,115,374 users.

#17 SPORTS : has average of 3,750,580 users.

#18 FAMILY : has average of 3,718,295 users.

#19 ART_AND_DESIGN : has average of 1,986,335 users.

#20 FOOD_AND_DRIN

# Conclusions:

## what we went through:

we wnt throw moplies applecations for two known markets google store and apple store; at first we had a raw data to work with we Revise and filter it and cleaned it from mistakes like bad entries and gaps and even multiple dublicates and isloated the non english apps, because the company we giving them the analysis results are interested only at making english apps.


## advices and tips:

1- making apps with game as a category is common between companies, but their's outcomes is significantly good as we saw in genres percentage table.

2- people now a days love both socialise and communicate in social network platforms.

3-media and photographic apps such ass video players and photo editors and even music players are well known between users in `both markets`.

4- the last advice is that making an app having fetures like socialize and media player will be  an good idea maybe the best idea on what we analyzed;
and it is
optional if the app has soem mini games to  entertain the bored audience while they exploring our app.